// @dcloudio/uni-app-vite/dist/nvue/plugin/transforms/transformText.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformText = void 0;
const uni_cli_shared_1 = require("@dcloudio/uni-cli-shared");
function isTextNode({ tag }) {
    return tag === 'text' || tag === 'u-text' || tag === 'button';
}
function isText(node) {
    const { type } = node;
    return (type === 2 /* NodeTypes.TEXT */ ||
        type === 12 /* NodeTypes.TEXT_CALL */ ||
        type === 5 /* NodeTypes.INTERPOLATION */ ||
        type === 8 /* NodeTypes.COMPOUND_EXPRESSION */);
}
const transformText = (node, _) => {
    if (!(0, uni_cli_shared_1.isElementNode)(node)) {
        return;
    }
    if (isTextNode(node)) {
        return;
    }
    const { children } = node;
    if (!children.length) {
        return;
    }
    children.forEach((child, index) => {
        if (isText(child)) {
            children.splice(index, 1, createText(node, child));
        }
    });
};
exports.transformText = transformText;
function createText(parent, node) {
    return {
        tag: 'u-text',
        type: 1 /* NodeTypes.ELEMENT */,
        tagType: 0 /* ElementTypes.ELEMENT */,
        props: [],
        isSelfClosing: false,
        children: [node],
        codegenNode: undefined,
        ns: parent.ns,
        loc: node.loc,
    };
}
